### PR TITLE
[theia-electron] Use electron-builder js api

### PR DESCRIPTION
**Pull request by marechal-p**
The CLI was marked as soon-to-be deprecated when using
`electron-builder`, so I converted the yaml configuration to a regular
js one.

Also makes the configuration use the electron executable downloaded by
`@theia/electron`, which can be patched to remove proprietary codecs.

Signed-off-by: Paul Maréchal <paul.marechal@ericsson.com>

### CODE REVIEW

**Comment by marechal-p** [comment]
cc @kittaakos I think you were the one to setup the Electron bundling?

I stayed as close as possible to what you did, tell me if I messed anything :)

**Comment by marechal-p** [comment]
linked https://github.com/theia-ide/theia/pull/5055

**Comment by marechal-p** [comment]
interesting? https://www.npmjs.com/package/electron-packager-plugin-non-proprietary-codecs-ffmpeg

**Comment by marcdumais-work** [comment]
@marechal-p I have added you as a collaborator on this repo. It doesn't change anything for this specific PR I think (the electron packaging app is not covered by CI), but in the future you can push your PRs branch directly to this repo and have the benefit of working CI.

**Comment by kittaakos** [comment]
> The CLI was marked as soon-to-be deprecated when

Where did you see this?

Can we use pure JS with `// @ts-check` for those two tiny modules?

File: theia-electron/electron-builder/src/options.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
+import { Configuration } from 'electron-builder';
+
+export function getTheiaBundleConfiguration(preview?: boolean): Configuration {
+    return {
+        electronDist: 'node_modules/electron/dist',
```
Comment (kittaakos) [inline theia-electron/electron-builder/src/options.ts]: Why is this required?

File: theia-electron/electron-builder/src/options.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+            'package.json'
+        ],
+        win: {
+            target: preview ? 'dir' : [
+                'nsis'
+            ]
+        },
+        mac: {
+            target: preview ? 'dir' : [
+                'dmg',
```
Comment (kittaakos) [inline theia-electron/electron-builder/src/options.ts]: [minor]: `,` but it is up to you.

**Comment by kittaakos** [comment]
I have tried it, it worked 👍 Thank you!

Before approving it; why do we need this change? It's nice we can build it with from JS insead of the CLI but what's the benefit of this change?

File: theia-electron/electron-builder/src/options.ts (hunk 1 above)
Comment (marechal-p) [reply theia-electron/electron-builder/src/options.ts]: `electronDist` tells `electron-builder` that it shouldn't pull binaries by itself, since we already have Electron installed inside `node_modules` thanks to `@theia/electron`. This distribution is also the one that is being patched by the scripts I added recently, so we really want to take full benefit from it.

**Comment by marechal-p** [comment]
> Before approving it; why do we need this change? It's nice we can build it with from JS insead of the CLI but what's the benefit of this change?

At first I hoped to use some plugin, but I realized it was for a different bundler :(

On the other hand I also have something outputted when packaging on master:

```sh
$ yarn package:preview
yarn run v1.16.0
$ electron-builder --dir
Configuring yargs through package.json is deprecated and will be removed in the next major release, please use the JS API instead.
Configuring yargs through package.json is deprecated and will be removed in the next major release, please use the JS API instead.
  • electron-builder version=20.40.2
[...]
```

**Comment by marechal-p** [comment]
But the two main advantages of this change are:

1. We use the patched Electron binaries pulled by `@theia/electron`.
2. We use the recommended JS API rather than the soon-to-be deprecated CLI one.

**Comment by kittaakos** [comment]
> 1. We use the patched Electron binaries pulled by `@theia/electron`.

OK, so this PR depends on https://github.com/theia-ide/theia/pull/5112, right?

> 2\. We use the recommended JS API rather than the soon-to-be deprecated CLI one.

You might be confused with this: https://github.com/electron-userland/electron-builder/pull/3852

It's not the `electron-builder` CLI that will be deprecated.

**Comment by kittaakos** [comment]
> OK, so this PR depends on [theia-ide/theia#5112](https://github.com/theia-ide/theia/pull/5112), right?

This part works great!

**Comment by marechal-p** [comment]
> [](https://github.com/electron-userland/electron-builder/pull/3852)
> 
> It's not the `electron-builder` CLI that will be deprecated.

Ah indeed, I opened another more minimal change: https://github.com/theia-ide/theia-apps/pull/180
